copy/copy.go

```diff
@@ -655,17 +725,122 @@ func (ic *imageCopier) copyLayer(ctx context.Context, srcInfo types.BlobInfo, po
+			// be a straight-forward mechanism to detect if another process
+			// is currently copying a specific layer.  However, with
+			// LockBlob() we are constrained by the file locks in
+			// containers/storage which cannot support try-lock semantics
+			// as fcntl(2) does not support that.
+			//
+			// The trick we are using here is to acquire the lock in a
+			// separate goroutine. The main routine first sleeps for
+			// the progressBarRefreshRate amount of time and will be
+			// signaled via a semaphore if and when the blob lock has
+			// been acquired.
+			blobLockSem := semaphore.NewWeighted(1)
+			blobLockSem.Acquire(ctx, 1)
+			var lockError error
+			go func() {
+				lockError = ic.c.dest.LockBlob(srcInfo)
+				blobLockSem.Release(1)
+			}()
+
+			copiedByAnotherProcess := false
+			if !blobLockSem.TryAcquire(1) {
+				// The initial sleep time is `progressBarRefreshRate`
+				// to make sure the bar is properly rendered before
+				// replacing it.
+				time.Sleep(progressBarRefreshRate)
+				if !blobLockSem.TryAcquire(1) {
+					logrus.Debugf("Blob %s is already being copied by another process", srcInfo.Digest)
+					copiedByAnotherProcess = true
+					// If we still don't have the lock, replace the
+					// initial progress bar with a new one
+					// indicating that the blob is copied by
+					// another process.
+					//
+					// Note that we call SetTotal() after the new
+					// bar is created to avoid potential flickering
+					// if the bar removal and replacement happens
+					// across refresh rates.
+					toReplaceBar := bar
+					bar = ic.c.createProgressBar(pool, srcInfo, "blob", "paused: being copied by another process", mpb.BarRemoveOnComplete(), mpb.BarReplaceOnComplete(bar))
+					toReplaceBar.SetTotal(0, true)
+					// block until we acquired the blob lock
+					blobLockSem.Acquire(ctx, 1)
+				}
+			}
+
+			if lockError != nil {
+				return false, types.BlobInfo{}, "", errors.Wrapf(err, "Error acquiring lock for blob %s", srcInfo.Digest)
+			}
+
+			// If the blob has not been copied by another process
+			// return immediately.
+			if !copiedByAnotherProcess {
+				logrus.Debugf("Blob %s is not being copied by another process", srcInfo.Digest)
+				return false, types.BlobInfo{}, "", nil
+			}
+
+			// If the blob has been copied by another process, it
+			// must be present in the in the destintation and be
+			// available for reuse.
+			reused, blobInfo, err = ic.c.dest.TryReusingBlob(ctx, srcInfo, ic.c.blobInfoCache, ic.canSubstituteBlobs)
+			if err != nil {
+				return false, types.BlobInfo{}, "", errors.Wrapf(err, "Error trying to reuse blob %s at destination", srcInfo.Digest)
+			}
+
+			if reused {
+				ic.c.dest.UnlockBlob(srcInfo)
+				logrus.Debugf("Blob %s has been successfully copied by another process", srcInfo.Digest)
+				toReplaceBar := bar
+				ic.c.createProgressBar(pool, srcInfo, "blob", "done: copied by another process", mpb.BarReplaceOnComplete(bar))
```

Comment:
Is the return value ever completed? If not, how does `pool.Wait` finish?

Reply:
It will complete. createProgressBar() will set the size to 0 if a filler is specified.